state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -809,6 +830,44 @@ pub mod pallet {
 			Self::deposit_event(Event::<T, I>::CcmBroadcastFailed { broadcast_id });
 			Ok(())
 		}
+
+		/// Called when an unfinalised deposit has been seen at the given address.
+		///
+		/// Requires `EnsureWitnessed` origin.
+		#[pallet::call_index(6)]
+		#[pallet::weight(Weight::zero())] // TODO JAMIE: Weight?
+		pub fn prewitness_deposits(
+			origin: OriginFor<T>,
+			deposit_witnesses: Vec<DepositWitness<T::TargetChain>>,
+			block_height: TargetChainBlockNumber<T, I>,
+		) -> DispatchResult {
+			T::EnsureWitnessed::ensure_origin(origin)?;
+
+			Self::deposit_event(Event::<T, I>::PrewitnessedDeposit {
+				deposit_witnesses: deposit_witnesses.clone(),
+				block_height,
+			});
+
+			for DepositWitness { ref deposit_address, asset, amount, .. } in deposit_witnesses {
+				let id = BoostableDepositIdCounter::<T, I>::mutate(|id| -> u64 {
+					*id = id.saturating_add(1);
+					*id
+				});
+
+				let deposit_channel_details = DepositChannelLookup::<T, I>::get(deposit_address)
+					.ok_or(Error::<T, I>::InvalidDepositAddress)?;
+
+				BoostableDeposits::<T, I>::insert(
```

Comment:
We're currently just inserting, but we also need to remove to make sure this storage doesn't expand forever. There are a few cases where it's removed:
1. It's executed (not part of this PR)
2. It's finalised (not part of this PR)
3. Nothing happens. This is possible if we get a pre-witnessed deposit that then never reaches finalisation (it's re-org'd out). In this case we want to clean up when the channel expires. We should do this as part of this PR. 

So, we just have to work out how to avoid a linear search, which will require a little bit of storage shifting. 

What we could do is this:
BoostableDeposits stored as a storage double map `Map<Address<Map<BoostableDepositId, Deposit>`. 

When we create the deposit, we also add to another storage item `BoostId -> DepositAddress`. Note, that the BoostId would still increment in the same way, on *every* deposit i.e. not just per channel.

Now, when a channel expires, we can just remove the `address` key. 

When we call `boost` extrinsic we can still use just `boost_id`, but in the code we would do a `id -> address` lookup, so then we can lookup the deposit. using `address`, then `boost_id`. 

Tagging @dandanlen here too for thoughts on above^

Reply:
Sounds good. To avoid the extra lookup, which would (I think?) require an extra map to store the correspondance between id and address, we could require boosters to provide the channel id or address with the deposit id. 

Ie. the booster would provide (effectively) a tuple of (channel_id, deposit_id) to identify the BoostableDeposit.

Reply:
Yeah avoided wanted to avoid that for simplicity for boosting - but perhaps it's an over-optimisation for now.

Reply:
Actually, I think (channel_id, deposit_id) is the only way to do it, because we have to clean up the ` BoostId -> DepositAddress` mapping too otherwise, which we would have to do with *another* storage item.

@GabrielBuragev tl;dr for you. Proposed change is that the `boost()` extrinsic would take `deposit_address, boost_id` rather than just `boost_id` as before. I don't think this should have an impact on you though?

Reply:
Yep, no impact on our end. Thank you for tagging me here!

Reply:
I would suggest (channel_id, deposit_id) because BoostableDeposit doesn't contain the address. Otherwise the booster will have to look up the address.

Reply:
> BoostableDeposit doesn't contain the address

I suggested above that it should include the address which would solve this